mu/modes/microbit.py

```diff
@@ -291,34 +312,86 @@ def flash(self):
                              self.user_defined_microbit_path))
         # Check the path and that it exists simply because the path maybe based
         # on stale data.
-        logger.debug('Path to micro:bit: {}'.format(path_to_microbit))
         if path_to_microbit and os.path.exists(path_to_microbit):
-            logger.debug('Flashing to device.')
+            force_flash = False  # If set to true, fully flash the device.
+            logger.info("Checking target device.")
+            # Get the version of MicroPython on the device.
+            try:
+                version_info = microfs.version()
+                logger.info(version_info)
+                board_version = version_info['version'].split('-', 1)[0]
+                board_version = board_version.replace('micro:bit v', '')
+                logger.info('Board MicroPython: {}'.format(board_version))
+                # If there's an older version of MicroPython on the device,
+                # update it with the one packaged with Mu.
+                if board_version < self.micropython_version:
+                    force_flash = True
+            except Exception:
+                # Could not get version of MicroPython. This means either the
+                # device has a really old version of MicroPython or is running
+                # something else. In any case, flash MicroPython onto the
+                # device.
+                logger.warn('Could not detect version of MicroPython.')
+                force_flash = True
             # Check use of custom runtime.
             rt_hex_path = self.editor.microbit_runtime.strip()
             message = _('Flashing "{}" onto the micro:bit.').format(tab.label)
             if (rt_hex_path and os.path.exists(rt_hex_path)):
                 message = message + _(" Runtime: {}").format(rt_hex_path)
+                force_flash = True  # Using a custom runtime, so flash it.
             else:
                 rt_hex_path = None
                 self.editor.microbit_runtime = ''
-            self.editor.show_status_message(message, 10)
-            self.set_buttons(flash=False)
-            self.flash_thread = DeviceFlasher([path_to_microbit],
-                                              python_script, rt_hex_path)
-            if sys.platform == 'win32':
-                # Windows blocks on write.
-                self.flash_thread.finished.connect(self.flash_finished)
+            # Check for use of user defined path (to save hex onto local
+            # file system.
+            if self.user_defined_microbit_path:
```

Comment:
How is this member variable is reset? I can only find something on line 324 of the original file (added a comment below to highlight it), and is only executed if `path_to_microbit` or `os.path.exists(path_to_microbit)` are `False`.

If a micro:bit is not plugged, Mu will ask the user to locate it (or any directory), and it will save that path into self.user_defined_microbit_path, if a micro:bit is then plugged, how does Mu that it needs to flash the micro:bit instead of the self.user_defined_microbit_path path?
Woudn't this always force a flash after the path is selected once?

Same goes for other uses of `if self.user_defined_microbit_path:`.